### PR TITLE
[plugin-lighthouse] Introduce Lighthouse plugin

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: docs/modules/plugins/nav.adoc [hunk 1]
```diff
@@ -49,3 +49,4 @@
 * xref:plugin-web-app.adoc[Web Application]
 * xref:plugin-web-app-to-rest-api.adoc[Web Application & REST API Integration]
 * xref:plugin-websocket.adoc[WebSockets]
+* xref:plugin-lighthouse.adoc[Lighthouse]
```
Comment (valfirst) [inline docs/modules/plugins/nav.adoc:52]: alphabetical order?

File: docs/modules/plugins/pages/plugin-lighthouse.adoc [hunk 2]
```diff
@@ -0,0 +1,61 @@
+= Lighthouse Plugin
+
+The plugin provides the ability to generate Lighthouse performance reports on the user experience of a page on both mobile and desktop devices, and provides suggestions on how that page may be improved.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-lighthouse.adoc:3]: need to add link to Lighthouse

File: docs/modules/plugins/pages/plugin-lighthouse.adoc [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+== Steps
+
+=== Perform Lighthouse scan
+
+Performs Lighthouse scan on the specified page and verifies performance result metrics.
+
+[source,gherkin]
+----
+When I perform Lighthouse scan on $strategy for `$url` page:$rules
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-lighthouse.adoc:36]: > Lighthouse scan on $strategy

it doesn't sound like English sentence, need to rename `$strategy` parameter

Comment (valfirst) [inline docs/modules/plugins/pages/plugin-lighthouse.adoc:36]: > $url

need to be more specific, like: `$webPageUrl`

File: docs/modules/plugins/pages/plugin-lighthouse.adoc [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+----
+
+* `$strategy` - The analysis strategy to use, either `desktop` or `mobile`.
+* `$url` - The page URL to perform scan on.
+* `$rules` - The metric verification rules:
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-lighthouse.adoc:41]:
```suggestion
* `$meticsValidations` - The metrics validations:
```

File: docs/modules/plugins/pages/plugin-lighthouse.adoc [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+** `$metric` - The Lighthouse metrics such as:
+*** `First Contentful Paint`
+*** `Largest Contentful Paint`
+*** `Total Blocking Time`
+*** `Cumulative Layout Shift`
+*** `Speed Index`
+** `$rule` - xref:parameters:comparison-rule.adoc[The comparison rule].
+** `$value` - The expected numeric metric value.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-lighthouse.adoc:49]: not clear what `numeric` is? is it integer/floating?

File: settings.gradle [hunk 6]
```diff
@@ -45,6 +45,7 @@ include 'vividus-plugin-excel'
 include 'vividus-plugin-html'
 include 'vividus-plugin-json'
 include 'vividus-plugin-lambdatest'
+include 'vividus-plugin-lighthouse'
```
Comment (valfirst) [inline settings.gradle:48]: need to add the plugin to docker image as well

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                .get("metrics").getDetails().get("items");
+        ArrayMap<String, BigDecimal> metrics = items.get(0);
+
+        softAssert.runIgnoringTestFailFast(() -> rules.forEach(rule ->
+        {
+            String metricKey = rule.getMetric();
+            String processedMetricKey = metricKey.replaceAll("\\s*", "");
+
+            Optional<String> metric = metrics.keySet().stream()
+                                                      .filter(m -> StringUtils.equalsIgnoreCase(m, processedMetricKey))
```
Comment (valfirst) [inline vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java:104]: why is `StringUtils` needed?

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+                                                      .findFirst();
+
+            if (metric.isEmpty())
```
Comment (valfirst) [inline vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java:107]: `ifPreserntOrElse` ?

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/model/PerformanceMetricRule.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+package org.vividus.lighthouse.model;
+
+import java.math.BigDecimal;
+
+import org.jbehave.core.annotations.AsParameters;
+import org.vividus.steps.ComparisonRule;
+
+@AsParameters
```
Comment (valfirst) [inline vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/model/PerformanceMetricRule.java:24]: just curious: is it required?

File: vividus-plugin-lighthouse/src/main/resources/org/vividus/lighthouse/lighthouse.ftl [hunk 10]
```diff
@@ -0,0 +1,283 @@
+<#noparse>
```
Comment (valfirst) [inline vividus-plugin-lighthouse/src/main/resources/org/vividus/lighthouse/lighthouse.ftl:1]: what is the source?

File: vividus-tests/src/main/resources/story/integration/LighthouseSteps.story [hunk 11]
```diff
@@ -0,0 +1,13 @@
+Description: Integration tests for LighthouseSteps class.
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/LighthouseSteps.story:1]: it can be omitted

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            {
+                softAssert.recordFailedAssertion(String.format("Unknown metric name '%s', available names are: %s",
+                        metricKey, String.join(", ", metrics.keySet())));
+                return;
+            }
+
+            softAssert.assertThat(String.format("The %s metric value", metricKey), metrics.get(metric.get()),
```
Comment (valfirst) [inline vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java:114]:
```suggestion
            softAssert.assertThat(String.format("%s metric", metricKey), metrics.get(metric.get()),
```

File: vividus-plugin-lighthouse/src/main/resources/org/vividus/lighthouse/lighthouse.ftl (hunk 10 above)
Comment (valfirst) [reply vividus-plugin-lighthouse/src/main/resources/org/vividus/lighthouse/lighthouse.ftl:1]: what is a source of the code from this FTL?

File: docs/modules/plugins/pages/plugin-lighthouse.adoc (hunk 5 above)
Comment (uarlouski) [reply docs/modules/plugins/pages/plugin-lighthouse.adoc:49]: numeric value refers to both integers and floating numbers

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/model/PerformanceMetricRule.java (hunk 9 above)
Comment (uarlouski) [reply vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/model/PerformanceMetricRule.java:24]: Caused by: org.jbehave.core.steps.ParameterConverters$ParameterConversionFailed: No parameter converter for java.util.List<org.vividus.lighthouse.model.PerformanceMetricRule>
        at org.jbehave.core.steps.ParameterConverters.convert(ParameterConverters.java:411) ~[jbehave-core-5.1.1-alpha.2.jar:?]
        at org.vividus.steps.ParameterConvertersDecorator.convertAdaptedValue(ParameterConvertersDecorator.java:81) ~[vividus-engine-0.5.11-SNAPSHOT.jar:0.5.11-SNAPSHOT]
        at org.vividus.steps.ParameterConvertersDecorator.convert(ParameterConvertersDecorator.java:54) ~[vividus-engine-0.5.11-SNAPSHOT.jar:0.5.11-SNAPSHOT]
        at org.jbehave.core.steps.StepCreator$ParametrisedStep.convertParameterValues(StepCreator.java:970) ~[jbehave-core-5.1.1-alpha.2.jar:?]
        at org.jbehave.core.steps.StepCreator$ParametrisedStep.parametriseStep(StepCreator.java:954) ~[jbehave-core-5.1.1-alpha.2.jar:?]
        at org.jbehave.core.steps.StepCreator$ParametrisedStep.perform(StepCreator.java:891) ~[jbehave-core-5.1.1-alpha.2.jar:?]

as I understand its used to determine whether to perform field mapping or to directly call converter

File: vividus-plugin-lighthouse/src/main/resources/org/vividus/lighthouse/lighthouse.ftl (hunk 10 above)
Comment (uarlouski) [reply vividus-plugin-lighthouse/src/main/resources/org/vividus/lighthouse/lighthouse.ftl:10]: added upgrade information

File: docs/modules/plugins/pages/plugin-lighthouse.adoc [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+|`lighthouse.application-name`
+|<string>
+|
+|The application name to be used in the https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent[User-Agent] header.
+
+|`lighthouse.access-token`
+|<string>
+|
+|The OAuth access token. Its recommended to specify the token if you plan to use the plugin in an automated way and making multiple queries per second.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-lighthouse.adoc:24]:
```suggestion
|The OAuth access token. Its recommended to specify the token if you plan to use the plugin in an automated way and make multiple queries per second.
```

File: docs/modules/plugins/pages/plugin-lighthouse.adoc [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+|===
+
+== Steps
+
+=== Perform Lighthouse scan
+
+Performs Lighthouse scan on the specified web page and verifies performance metrics results.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-lighthouse.adoc:32]:
```suggestion
Performs Lighthouse scan of the specified web page and validates performance metrics against expected thresholds.
```

File: docs/modules/plugins/pages/plugin-lighthouse.adoc [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+* `$webPageUrl` - The page URL to perform scan on.
+* `$meticsValidations` - The metric validations:
+** `$metric` - The Lighthouse metric name such as:
+*** `First Contentful Paint`
+*** `Largest Contentful Paint`
+*** `Total Blocking Time`
+*** `Cumulative Layout Shift`
+*** `Speed Index`
+** `$rule` - xref:parameters:comparison-rule.adoc[The comparison rule].
+** `$value` - The expected integer or floating (e.g. `0.35`) number.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-lighthouse.adoc:52]:
```suggestion
** `$threshold` - The expected integer or floating (e.g. `0.35`) number.
```

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+                    req.setWriteTimeout(DEFAULT_TIMEOUT);
+                }
+            )
+            .setApplicationName(applicationName)
+            .build();
+        this.accessToken = accessToken;
+    }
+
+    /**
+     * Performs Lighthouse scan on the specified page and verifies performance result metrics.
```
Comment (valfirst) [inline vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java:70]: do not forget to adjust javadocs according to comments in regular docs

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+        for (String strategy : scanType.getStrategies())
+        {
+            LighthouseResultV5 result = executePagespeed(webPageUrl, strategy);
+
+            String resultAsString = jsonFactory.toString(result);
+
+            attachmentPublisher.publishAttachment("/org/vividus/lighthouse/lighthouse.ftl",
+                    Map.of("result", resultAsString),
+                    String.format("Lighthouse %s report for page: %s", strategy, webPageUrl));
```
Comment (valfirst) [inline vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/LighthouseSteps.java:92]:
```suggestion
                    String.format("Lighthouse %s scan report for page: %s", strategy, webPageUrl));
```

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/model/PerformanceMetricRule.java (hunk 9 above)
Comment (valfirst) [reply vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/model/PerformanceMetricRule.java:24]: we can think of defaulting to field mapping if parameter converter is not available

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/model/PerformanceMetricRule.java (hunk 9 above)
Comment (uarlouski) [reply vividus-plugin-lighthouse/src/main/java/org/vividus/lighthouse/model/PerformanceMetricRule.java:24]: https://jbehave.atlassian.net/browse/JBEHAVE-1581

**Review by valfirst — APPROVED** [review]

**Pull request merged**